### PR TITLE
Make an ember-cli addon.

### LINKED ISSUE #79: Move to node
This library relies on the ember-dev Ruby stuff. I know that @rwjblue has the knowledge in his head about how to transition properly. I can do it, just point me in the right direction

### LINKED ISSUE #102: This doesn't seem to be published on npm

### LINKED ISSUE #106: Don't Rely On Globals
The resolver currently depends on an Ember global. It would be great if it did not, so ordering does not matter during concatenation.

### PR DESCRIPTION
**Pull request by rwjblue**
- [x] Migrate tests.
- [x] Avoid using shared module registry directly.  Change to `has`/`get` methods.
- [x] Fix ember-try config (ensure Ember 1.10, 1.11, 1.12, 1.13, 2.0 are all tested).
- [x] Submit PR to Ember CLI (https://github.com/ember-cli/ember-cli/pull/4777).
  - [x] Remove usage of `ember/resolver` in favor of `ember-resolver` (app.js & tests/helpers/resolver.js).
  - [x] Remove `ember-resolver.js` from `vendorFiles` with backwards compatibility.
- [x] Confirm all works properly with ember-inspector (cc @teddyzeenny)
- [x] Update `moduleNameLookupPatterns` and `findModuleName` to avoid `Ember.A` usage.
- [x] Confirm next version number should be 1.0.0
- [x] Get permission to publish `ember-resolver` (see https://github.com/seeden/ember-require-resolver/issues/2 for details).

---

Fixes #79
Fixes #102
Fixes #106

### CODE REVIEW

File: addon/utils/module-entries.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/*globals requirejs */
+
+if (typeof requirejs.entries === 'undefined') {
+  requirejs.entries = requirejs._eak_seen;
+}
+
+export default requirejs.entries;
```
Comment (stefanpenner) [inline addon/utils/module-entries.js]: i would prefer if we didn't share the `entries` directory, rather exposed accessors. Sharing the raw structure gives me the heebee jeebees.

* `has`
* `get`

File: addon/resolver.js [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+      this.mainModuleName,
+      this.defaultModuleName
+    ]);
+  }),
+
+  findModuleName: function(parsedName, loggingDisabled){
+    var self = this;
+    var moduleName;
+
+    this.get('moduleNameLookupPatterns').find(function(item) {
```
Comment (stefanpenner) [inline addon/resolver.js]: isn't needed for this PR. But I would prefer if we didn't rely on prototype extensions here. Instead just used a for-loop, as this little loop does get called quite often during resolution

File: addon/utils/module-entries.js (hunk 1 above)
Comment (rwjblue) [inline addon/utils/module-entries.js]: Seems good.  I'll update.

**Comment by stefanpenner** [comment]
:+1:

File: addon/resolver.js (hunk 2 above)
Comment (rwjblue) [inline addon/resolver.js]: Sounds good. Added to checklist here.

**Comment by stefanpenner** [comment]
cc @chadhietala

File: addon/utils/module-entries.js (hunk 1 above)
Comment (rwjblue) [inline addon/utils/module-entries.js]: Addressed by 5b09ec16accec0625327bff0114c52b31d16b80a.

File: addon/resolver.js (hunk 2 above)
Comment (rwjblue) [inline addon/resolver.js]: Array prototype extensions fixed in 4c1463c.

**Comment by quaertym** [comment]
:+1: , one less bower dependency :beers:

**Comment by chadhietala** [comment]
Seems good. So people will need to import `ember-resolver` instead of `ember/resolver`? Did we ever decide if the `@ember` scope was a viable option?

**Comment by rwjblue** [comment]
> So people will need to import ember-resolver instead of ember/resolver? 

Yep, along with some changes with vendorFiles in the build process. I'll have a PR updating ember-cli as part of the checklist at the top. 

> Did we ever decide if the @ember scope was a viable option?

I think we are generally in favor, but it is not finalized. I would prefer to not block this PR on that decision.

**Comment by rwjblue** [comment]
I made this compatible with Ember CLI 0.2.0+ (it removes the legacy `bower_components/ember-resolver/dist/ember-resolver.js` entry in `vendorFiles`) in b794a60.

**Comment by rwjblue** [comment]
Apparently, `ember-resolver` was previously registered on NPM (for a different thing that has subsequently changed its name).  I opened https://github.com/seeden/ember-require-resolver/issues/2 to ask if they can add us as owners to `ember-resolver`.  For now (to continue forward), I am going to publish as `ember-cli-resolver`.

**Comment by rwjblue** [comment]
OK, original ember-resolver was unpublished, but we can't use 1.0.0 because it had a 1.x series.  Switching to 2.0.0 (which seems "locksteppy" anyways).

**Comment by stefanpenner** [comment]
> Did we ever decide if the @ember scope was a viable option?

it is a viable option (in-fact i own the scope)

> I think we are generally in favor, but it is not finalized. I would prefer to not block this PR on that decision.

lets sync up about this, I would love to `finalize` this sooner rather then later. For now `ember-resolver` is fine, the transition to `@ember/resolver` wont be care painful.

**Comment by rwjblue** [comment]
> lets sync up about this

Sounds good. I think my main point is that I don't want to block current progress.  I'm happy to reconsider if you feel strongly otherwise (but it sounds like you are fine moving forward for now).

**Comment by stefanpenner** [comment]
> if you feel strongly otherwise (but it sounds like you are fine moving forward for now).

Yup, keep moving forward.

**Comment by stefanpenner** [comment]
created an issue to track scope related thoughts https://github.com/ember-cli/ember-resolver/issues/114

File: addon/container-debug-adapter.js [hunk 3]
```diff
@@ -58,7 +73,16 @@ define("ember/container-debug-adapter",
...
     _getEntries: function() {
-      return requirejs.entries;
+      // deprecate and confirm with @teddyzeenny that this is not needed by the inspector
+      // remove once confirmed
+
+      Ember.deprecate(
+        'Usage of ContainerDebugAdapter#_getEntries is deprecated and will be removed.',
+        false,
+        { id: 'ember-resolver.container-debug-adapter._getEntries', until: '1.0.0' }
+      );
```
Comment (teddyzeenny) [inline addon/container-debug-adapter.js]: No need to deprecate, this was private and only used internally.

**Comment by teddyzeenny** [comment]
Looks good on the inspector side :+1: 
`_getEntries` can be removed.

**Comment by rwjblue** [comment]
Thanks for checking @teddyzeenny!

File: addon/container-debug-adapter.js (hunk 3 above)
Comment (rwjblue) [inline addon/container-debug-adapter.js]: Awesome, removed now.

**Pull request merged**